### ISSUE TITLE
swupd is unable to downgrade from version

### ISSUE THREAD

**Issue by dborquez** [clearlinux/swupd-client#238]
In order to downgrade CLR from version: 17230 to version 16860 I executed:
* swupd verify -f -m 16860

* As result swupd reported this error:

```
...
Hash mismatch for file: /usr/share/vulkan/icd.d/intel_icd.x86_64.json
        fixed
Inspected 232925 files
  8231 files were missing
    8231 of 8231 missing files were replaced
    0 of 8231 missing files were not replaced
  7915 files did not match
    7915 of 7915 files were fixed
    0 of 7915 files were not fixed
  0 files found which should be deleted
Calling post-update helper scripts.
sh: relocation error: /usr/lib64/haswell/libc.so.6: symbol __tunable_get_val, version GLIBC_PRIVATE not defined in file ld-linux-x86-64.so.2 with link time reference
sh: relocation error: /usr/lib64/haswell/libc.so.6: symbol __tunable_get_val, version GLIBC_PRIVATE not defined in file ld-linux-x86-64.so.2 with link time reference
sh: relocation error: /usr/lib64/haswell/libc.so.6: symbol __tunable_get_val, version GLIBC_PRIVATE not defined in file ld-linux-x86-64.so.2 with link time reference
Fix successful
```

* After that Clear Linux was broken:

```
# cat /usr/lib/os-release
cat: relocation error: /usr/lib64/haswell/libc.so.6: symbol __tunable_get_val, version GLIBC_PRIVATE not defined in file ld-linux-x86-64.so.2 with link time reference
```

**Comment by phmccarty** [comment]
Please post the list of bundles you have installed.

**Comment by dborquez** [comment]
It just have this set of bundles:
- openssh-server
- clr-devops

**Comment by matthewrsj** [comment]
ok, I was able to reproduce this using the 17230 live image from https://download.clearlinux.org/releases/17230/ 

I created a gist of the full output here: https://gist.github.com/matthewrsj/b25f15da14768b97c63d4706ebaee3e0

I will be checking if this error exists when going the other direction, 16860 -> 17230.

@dborquez to fix your current system, take the following steps:
- Download the [clear linux installer image](https://download.clearlinux.org/current/clear-17260-installer.img.xz) to a usb drive.
- Boot the installer image on your borked system
  * In "Step 1 of 6 (Network Requirements)" verify that a connection the the update server has been `established`.
  * In "Step 2 of 6 (Choose Action)" select `Repair`. This should repair your system to the version it is currently at, 16860.
- Reboot into your host system and verify everything is working properly.

**Comment by matthewrsj** [comment]
Interestingly, I can only reproduce this error in the official images released since 16860. When I create an image myself using our [installer](https://github.com/bryteise/ister) I am able to freely fix back to 16860 with zero issues.

Perhaps this is a content issue with the official release? I'll investigate the file list for each image and see if there is a difference.

**Comment by phmccarty** [comment]
@dborquez Another question. Are you running swupd on a bare metal install of Clear Linux, in QEMU, or other?

**Comment by dborquez** [comment]
#matthewrsj , I rebuild the system, so I was unable to test the method you suggested.

**Comment by phmccarty** [comment]
I am able to reproduce this issue running a Clear Linux KVM image with QEMU on a Clear Linux host. A more minimal test is downgrading from version 16890 to 16860.

**Comment by phmccarty** [comment]
I also discovered what is the likely root cause. Details follow.

Background on the stream of changes introduced to Clear Linux: The glibc package was upgraded from 2.25 to 2.26 in build 16870, and the AVX2-optimized glibc (libc.so.6 symlink + libc-2.26.so) was added to a few bundles in 16890. In this scenario, if the AVX2 lib + lib symlink is present on the filesystem, the dynamic linker will use it if it can. Else, it will use the stock glibc library. So, in build 16870, the stock library is used, because the 16870 is not present in any bundle. In build 16890, the AVX2 lib is used, assuming the underlying system supports AVX2.

From the SWUPD side: Downgrading a system with `verify -f -m VER` from build 16890 (or newer) back to 16860 (or older) will *not* delete the AVX2 lib+symlink added in 16890, because it was not present in any manifests prior to 16890 and therefore SWUPD had no knowledge that the files were present and should be deleted -- `verify` only has knowledge about the current version, or the version passed to `-m`. Not deleting the files triggered the system corruption, because the AVX2 lib still installed on the system is from glibc 2.26, and the dynamic linker was just downgraded to glibc 2.25; the 2.25 dynamic linker cannot load the 2.26 lib.

Workaround: `rm /usr/lib64/haswell/libc-2.26.so /usr/lib64/haswell/libc.so.6` prior to the running the `verify -f -m VER` operation when downgrading a system in the version ranges described above.

**Comment by phmccarty** [comment]
As a possible solution for SWUPD to more gracefully handle this case, `verify` could set the `--picky` option in case the user is downgrading the system. This change of mode of operation would result in files under `/usr` being cleaned up if they are not listed in SWUPD manifests, addressing this particular bug.

Another option would be to teach `verify` to compare the _current version_ with the version passed to `-m` so that it could calculate precisely what it needs to delete during a downgrade. If the current version equals the `-m` version (or no `-m` option is passed), then there would be no change of behavior.

**Comment by icarus-sparry** [comment]
Perhaps if "-m" is specified (and is different to the current version) then we should abort unless either '--force' or '--picky' is specified, with a warning message saying one or the other (or both) is needed?

**Comment by matthewrsj** [comment]
The suggestion by @icarus-sparry was implemented in #242.
